zerver/views/message_fetch.py

```diff
@@ -169,74 +170,93 @@ def get_messages_backend(
+        # when messages are deleted in between.  We set up
+        # repeatable-read isolation for this transaction, so that we
+        # prevent both phantom reads and non-repeatable reads.
+        #
+        # In a read-only repeatable-read transaction, it is not
+        # possible to encounter deadlocks or need retries due to
+        # serialization errors.
+        cursor = connection.cursor()
+        cursor.exectute("SET TRANSACTION ISOLATION LEVEL REPEATABLE READ READ ONLY")
+
+        query_info = fetch_messages(
+            narrow=narrow,
+            user_profile=user_profile,
+            realm=realm,
+            is_web_public_query=is_web_public_query,
+            anchor=anchor,
+            include_anchor=include_anchor,
+            num_before=num_before,
+            num_after=num_after,
+        )
 
-        for message_id in message_ids:
-            if message_id not in user_message_flags:
-                user_message_flags[message_id] = ["read", "historical"]
-    else:
-        for row in rows:
-            message_id = row[0]
-            flags = row[1]
-            user_message_flags[message_id] = UserMessage.flags_list_for_flags(flags)
-            message_ids.append(message_id)
-
-    search_fields: Dict[int, Dict[str, str]] = {}
-    if is_search:
-        for row in rows:
-            message_id = row[0]
-            (topic_name, rendered_content, content_matches, topic_matches) = row[-4:]
-            search_fields[message_id] = get_search_fields(
-                rendered_content, topic_name, content_matches, topic_matches
+        anchor = query_info.anchor
+        include_history = query_info.include_history
+        is_search = query_info.is_search
+        rows = query_info.rows
+
+        # The following is a little messy, but ensures that the code paths
+        # are similar regardless of the value of include_history.  The
+        # 'user_messages' dictionary maps each message to the user's
+        # UserMessage object for that message, which we will attach to the
+        # rendered message dict before returning it.  We attempt to
+        # bulk-fetch rendered message dicts from remote cache using the
+        # 'messages' list.
+        message_ids: List[int] = []
+        user_message_flags: Dict[int, List[str]] = {}
+        if is_web_public_query:
+            # For spectators, we treat all historical messages as read.
+            for row in rows:
+                message_id = row[0]
+                message_ids.append(message_id)
+                user_message_flags[message_id] = ["read"]
+        elif include_history:
+            assert user_profile is not None
+            message_ids = [row[0] for row in rows]
+
+            # TODO: This could be done with an outer join instead of two queries
+            um_rows = UserMessage.objects.filter(
+                user_profile=user_profile, message_id__in=message_ids
             )
```

Comment:
Would addressing the `TODO` here be an alternative solution?

Reply:
Hmm, answer is no, but this should probably be using `UserMessageLite` or an `values` query to avoid doing all the lazy evaluation setup for the `.message.` system.

Reply:
I'm really not clear what lazy evaluation setup you're referencing.

And yeah, this query is kinda-irrelevant.  Making it an outer join would be pretty inefficient, and doesn't change the fact that the possible matching messages could change between the `fetch_messages` choosing some message-ids and the `message_for_ids` trying to look them up.

If anything, this is a _further_ argument in terms of wanting repeatable-read in here.